### PR TITLE
Feature: Implement SocketIO for calculate roll and bet results.

### PR DESCRIPTION
**Pull request by nddmanh**

### CODE REVIEW

File: services/socket.js [hunk 1]
```diff
@@ -1,18 +1,24 @@
...
+const jwt = require('jsonwebtoken');
 
 const activateSocket = (io) => {
   let timeLeft = TIME_LEFT_DEFAULT;
   let timeInterval;
   let userConnectedCounter = 0;
+  let userId;
 
-  const countdown = () => {
+  const countdown = async () => {
```
Comment (lethemanh) [inline services/socket.js:14]: Why you don't move ```countdown``` to another file than import to ```socker.js```, I think this function has separately logic

File: services/socket.js [hunk 2]
```diff
@@ -1,18 +1,24 @@
...
     if (timeLeft <= 0) {
       clearInterval(timeInterval);
       io.emit('processing-result');
       // Server roll result
-      rollResult();
+      await rollResult();
+
+      // Return result to player
+      const returnResult = await calculateResult(userId);
```
Comment (lethemanh) [inline services/socket.js]: ```returnResult``` can be undefined, so you still emit socket even if ```returnResult = undefined```? I think when this case hapend you have to return error for client.

File: helper/calculateResult.js [hunk 3]
```diff
@@ -22,14 +28,21 @@ const calculateResult = async (user) => {
     await userRepository.update(user._id, user); // Update player's money
   } 
 
+  let returnResult = {
```
Comment (lethemanh) [inline helper/calculateResult.js]:
```suggestion
  const returnResult = {
```

File: services/socket.js [hunk 4]
```diff
@@ -1,18 +1,24 @@
...
+  let userId;
 
-  const countdown = () => {
+  const countdown = async () => {
     if (timeLeft <= 0) {
       clearInterval(timeInterval);
       io.emit('processing-result');
       // Server roll result
-      rollResult();
+      await rollResult();
```
Comment (lethemanh) [inline services/socket.js:19]: Why you have ```await rollResult``` and don't have any returned value?

Comment (nddmanh) [reply services/socket.js:19]: i call functionn rollResult and it return value at that fuction.

**Review by lethemanh — APPROVED** [review]

File: helper/calculateEachUser.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+const calculateResult = require('./calculateResult');
+
+const calculateEachUser = (userIds) => {
+  return new Promise( async (resolve, reject) => {
+    try {
+      let result = userIds.map( async (userId) => {
+        return await calculateResult(userId);
+      });
+      resolve(await Promise.all(result));
```
Comment (ppppbn) [inline helper/calculateEachUser.js]: This is not how you handle an array of promises. Instead
`const promises = userIds.map(userId => calculateResult(userId))`
`return Promise.all(promises)`

File: helper/calculateEachUser.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+const calculateResult = require('./calculateResult');
+
+const calculateEachUser = (userIds) => {
```
Comment (ppppbn) [inline helper/calculateEachUser.js]: `calculateResults` or `calculateMultipleResults`

File: helper/calculateResult.js [hunk 7]
```diff
@@ -4,9 +4,20 @@ const betRepository = require('../modules/bets/bet.repository');
...
-  const bet = await betRepository.findBetProcessing();
+  const bet = await betRepository.findBetProcessing(userId);
+  let returnResult;
+
+  if (!bet) {
+    returnResult = {
+      newBalance: user.balance,
+      rollResult: roll.rollResult
+    };
+    await rollRepository.deleteRollProcessing();
```
Comment (ppppbn) [inline helper/calculateResult.js]: If you delete the processing roll here, other users wont have that roll to calculate the result

File: helper/calculateEachUser.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+  return new Promise( async (resolve, reject) => {
+    try {
+      let result = userIds.map( async (userId) => {
```
Comment (lethemanh) [inline helper/calculateEachUser.js]:
```suggestion
      let result = userIds.map(async (userId) => {
```

File: helper/calculateResult.js [hunk 9]
```diff
@@ -4,9 +4,20 @@ const betRepository = require('../modules/bets/bet.repository');
...
+    return returnResult;
+  }
   // Caculate duplicate coefficient
   let coefficient = 0;
   roll.rollResult.forEach( item => {
```
Comment (lethemanh) [inline helper/calculateResult.js:26]: Dangerous! What if ```roll``` is null or undefined?

File: helper/calculateResult.js [hunk 10]
```diff
@@ -22,14 +33,21 @@ const calculateResult = async (user) => {
...
   // Change status
   bet.status = STATUS.DONE;
   roll.status = STATUS.DONE;
 
   // Update Data
-  await rollRepository.update(Roll._id, Roll); 
-  await betRepository.update(Bet._id, Bet); 
-  return user.balance;
+  await rollRepository.update(roll._id, roll); 
+  await betRepository.update(bet._id, bet);
```
Comment (lethemanh) [inline helper/calculateResult.js:48]: Does update bet depend on update roll? If not you can do as suggestion
```await Promise.all([rollRepository.update(roll._id, roll), betRepository.update(bet._id, bet)])```

Comment (ppppbn) [reply helper/calculateResult.js:48]: Not clear and unnecessary

**Comment by ppppbn** [comment]
Also please discard your approval first before adding comments @lethemanh

File: helper/calculateMultipleResults.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    const roll = await rollRepository.findRollProcessing();
+    const countBetsProcessing = await betRepository.countBetsProcessing();
+    
+    const promises = userIds.map(userId => calculateEachUser(userId, roll));
+    let result = await Promise.all(promises);    
+    handleRollAfterEachTurn(countBetsProcessing, roll);
+
+    return result;
+  } catch (error) {
+    console.log(error);
```
Comment (lethemanh) [inline helper/calculateMultipleResults.js]: Remove console.log

File: helper/calculateEachUser.js [hunk 12]
```diff
@@ -1,12 +1,20 @@
...
+  if (!bet) {
+    returnResult = {
+      newBalance: user.balance,
+      rollResult: roll.rollResult
+    };
+    return returnResult;
+  }
   // Caculate duplicate coefficient
   let coefficient = 0;
   roll.rollResult.forEach( item => {
```
Comment (lethemanh) [inline helper/calculateEachUser.js:26]: If ```roll``` is null there will be an error, check ```roll``` not null or undefined first

**Review by lethemanh — APPROVED** [review]

File: helper/handleRoll.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
+const rollRepository = require('../modules/rolls/roll.repository');
+const { STATUS } = require('../constants/status');
+
+const handleRollAfterEachTurn = async (countBetsProcessing, roll) => {
+  if (countBetsProcessing === 0) {
+    await rollRepository.deleteRollProcessing();
+  } else {
+    roll.status = STATUS.DONE;
+    await rollRepository.update(roll._id, roll);
+  }
```
Comment (ppppbn) [inline helper/handleRoll.js:10]: Please explain the logic of this function

Comment (nddmanh) [reply helper/handleRoll.js:10]: In the next turns, we will find roll PROCESSING to handle. So, when end the turn, server will change status of roll from PROCESSING to DONE.

File: modules/users/user.router.js [hunk 14]
```diff
@@ -6,6 +6,7 @@ router.get('/profile', async function (req, res) {
   try {
     const data = await userService.getProfile(req.user);
     res.status(200).json({data: {
+      _id: data ? data._id : null,
       username: data ? data.username : null,
       email: data ? data.email : null,
       phoneNumber: data ? data.phoneNumber : null,
```
Comment (ppppbn) [inline modules/users/user.router.js]: Why do you have to check for `data` here for every field?

File: services/socket.js [hunk 15]
```diff
@@ -1,36 +1,75 @@
...
+          role: data.role
+        };
+        return next();
+      }
+      return next(new Error("Not authenticated!"));
+    } catch (err) {
+      return next(err);
+    }
+  }).on('connection', (socket) => { 
+    userIds.push(socket.user._id);
```
Comment (ppppbn) [inline services/socket.js]: Should filter before push, to prevent duplicated `_id`

File: helper/calculateEachUser.js [hunk 16]
```diff
@@ -1,12 +1,23 @@
...
-const calculateResult = async (user) => {
-  const roll = await rollRepository.findRollProcessing();
-  const bet = await betRepository.findBetProcessing();
+const calculateEachUser = async (userId, roll) => {
+  const user = await userRepository.findById(userId);
+  const bet = await betRepository.findBetProcessing(userId);
+  let returnResult;
+
+  if (!roll) {
+    return false;
```
Comment (ppppbn) [inline helper/calculateEachUser.js]: Return `false` here will destroy the data structure returned from the function. Better handle it just like the case where there is no `bet` as stated below

**Review by huyanh7796 — APPROVED** [review]

**Pull request merged**